financial-templates-lib/TokenBalanceClient.js

```diff
@@ -0,0 +1,135 @@
+const { delay } = require("./delay");
+
+// A thick client for getting information about an different token Balance from the chain.
+// This client is kept separate from the other clients to only store token balances for a given EMP.
+// After a balance is requested for a given wallet address that wallet is registered within a local
+// array of addresses that the client monitors. This lets bots that implement the client retrieve the
+// latest available data from the last update synchronously.
+class TokenBalanceClient {
+  constructor(logger, ERC20abi, web3, collateralTokenAddress, syntheticTokenAddress, updateThreshold = 60) {
+    this.logger = logger;
+    this.updateThreshold = updateThreshold;
+    this.lastUpdateTimestamp;
+
+    this.web3 = web3;
+
+    // Token contracts
+    this.collateralToken = new web3.eth.Contract(ERC20abi, collateralTokenAddress);
+    this.syntheticToken = new web3.eth.Contract(ERC20abi, syntheticTokenAddress);
+
+    // Token balances to enable synchronous return of the latest token ballance cashed in the client.
+    this.tokenBalances = { collateralBalances: {}, syntheticBalances: {}, etherBalances: {} };
+
+    // Array of balances to monitor. Updated when a new addresses balance is requested
+    this.accountMonitorList = [];
+  }
+
+  // Calls _update unless it was recently called, as determined by this.updateThreshold.
+  update = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    if (currentTime < this.lastUpdateTimestamp + this.updateThreshold) {
+      this.logger.debug({
+        at: "TokenBalanceClient",
+        message: "Token Balances update skipped",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTimestamp,
+        timeRemainingUntilUpdate: this.lastUpdateTimestamp + this.updateThreshold - currentTime
+      });
+      return;
+    } else {
+      await this._update();
+      this.lastUpdateTimestamp = currentTime;
+      this.logger.debug({
+        at: "TokenBalanceClient",
+        message: "Token Balances updated",
+        lastUpdateTimestamp: this.lastUpdateTimestamp
+      });
+    }
+  };
+
+  // Force call of _update, designed to be called by downstream caller that knowingly updated the Token Balances.
+  forceUpdate = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    await this._update();
+    this.lastUpdateTimestamp = currentTime;
+    this.logger.debug({
+      at: "TokenBalanceClient",
+      message: "Token Balances force updated",
+      lastUpdateTimestamp: this.lastUpdateTimestamp
+    });
+  };
```

Comment:
This logic is in all the other clients, too. I thought you had suggested moving all of this to a base class? (Same for `poll()` below)

Reply:
ye, this still needs to be done. It will be a separate PR as it'll change all the clients. It hasn't been done yet.